// Controllers/companies-controller.js
var dataContext = require('../../models/server-context.js');
var Q = require("q");

module.exports = {
    getAll: function () {
        var deferred = Q.defer();

        dataContext().then(function (db) {
            db.models.company.find({}, function (error, results) {
                if (error) {
                    deferred.reject(error);
                }

                if (results && results.length > 0) {
                    deferred.resolve(results);
                }
            });
        }, function (error) {
            deferred.reject(error);
        });

        return deferred.promise;
    },
    getResourcesByCompany: function (companyId) {
        var deferred = Q.defer();

        dataContext().then(function (db) {
            db.models.resource
                .findByCompany({ companyId: companyId }, function (error, results) {
                    if (error) {
                        deferred.reject(error);
                    }

                    deferred.resolve(results);
                });
        }, function (error) {
            deferred.reject(error);
        });

        return deferred.promise;
    },
    getResourceDataByCompany: function (companyId, resourceId) {
        var deferred = Q.defer();

        dataContext().then(function (db) {
            db.models.resource
                .findByCompany({ companyId: companyId }, function (error, results) {
                    if (error) {
                        deferred.reject(error);
                    }

                    deferred.resolve(results.find(function (res) {
                        return res.resourceId == resourceId
                    }));
                });
        }, function (error) {
            deferred.reject(error);
        });

        return deferred.promise;
    },
    getEventsByCompany: function (companyId) {
        var deferred = Q.defer();

        dataContext().then(function (db) {
            db.driver.execQuery(
                "SELECT * FROM event e INNER JOIN resource r on e.resource_resourceId = r.resourceId WHERE r.company_companyId = ?", [companyId],
                function (err, data) {
                    if (err) {
                        deferred.reject(err);
                    }

                    deferred.resolve(data);
                });
        }, function (error) {
            deferred.reject(error);
        });

        return deferred.promise;
    },
    getEventDataByCompany: function (companyId, eventId) {
        var deferred = Q.defer();

        dataContext().then(function (db) {
            db.driver.execQuery(
                "SELECT * FROM event e INNER JOIN resource r on e.resource_resourceId = r.resourceId WHERE r.company_companyId = ? AND e.eventId = ?",
                [companyId, eventId],
                function (err, data) {
                    if (err) {
                        deferred.reject(err);
                    }

                    deferred.resolve(data[0]);
                });
        }, function (error) {
            deferred.reject(error);
        });

        return deferred.promise;
    }
};
